### PR TITLE
Core & Internals: Filtering engine for advanced metadata search; Fix #3750

### LINKED ISSUE #3750: Support DID metadata queries with inequality operators:
Motivation
----------
This is a request for support for did metadata queries using inequalities like:
```
rucio list-dids-by-meta --scope blah --key>some_value
```
where `key` is something like a `gps-start-time` with a numerical value (e.g. 1276523729) associated with the DID.

The particular use case here is for LIGO/Virgo where our rucio-managed files contain time-series data and are associated with some time segment.  

Modification
------------
Introduce inequalities and integer/float values to the metadata query in [did.py](https://github.com/rucio/rucio/blob/master/lib/rucio/core/did.py#L1477)

### PR DESCRIPTION
**Pull request by robbarnsley**
Refactoring of https://github.com/rucio/rucio/pull/4139

Note that filtering engine == inequality engine.

Some notable changes from 4139 and this PR:

- <filters> variable kept as single type as it progresses through client/rest/api workflow resulting in fewer changes required to these interfaces
- parsing of filter string moved to common/utils, as per previous workflow
- parsed <filters> formatted as "keyword.operator", complementing current syntax e.g. "length.gte"
- keyword sanity checking moved entirely to inside filter engine
- sqla queries generated by filter engine now compounded to single query
- != operator added with wildcard support (NOT LIKE)
- added logic to filter duplication of dids resulting from non-mutual result sets 
- fixed issue with "length" keyword backwards compatibility

### CODE REVIEW

**Comment by robbarnsley** [comment]
Just to note, this PR only improves searching for `rucio list-dids`, not extended, and as such is only for the hardcoded DID columns.

If I can get this reviewed, I would be looking next to expand the functionality to include JSON metadata.

File: bin/rucio [hunk 1]
```diff
@@ -2040,7 +2052,7 @@ You can filter by key/value, e.g.::
 
     list_parser.set_defaults(function=list_dids)
     list_parser.add_argument('--recursive', dest='recursive', action='store_true', default=False, help='List data identifiers recursively.')
-    list_parser.add_argument('--filter', dest='filter', action='store', help='Filter arguments in form `key=value,another_key>=next_value`. Valid keys are type and length. Use `type` to specify a DID type to search for. Valid values are `ALL`, `COLLECTION`, `CONTAINER`, `DATASET`, `FILE`. Use `length` to filter for an amount of corresponding Data IDentifiers. Valid operators are `=`, `>`, `>=`, `<`, `<=`.')  # noqa: E501
+    list_parser.add_argument('--filter', dest='filter', action='store', help='Filter arguments in form supported by the inequality-engine.')
```
Comment (mlassnig) [inline bin/rucio]: Maybe instead just write what the form is?

File: lib/rucio/common/exception.py [hunk 2]
```diff
@@ -1023,5 +1023,25 @@ class UnsupportedAccountName(RucioException):
...
         self.error_code = 97
+
+
+class DuplicateCriteriaInDIDFilter(RucioException):
+    """
+    Duplicate criteria found in DID filter.
+    """
+    def __init__(self, *args, **kwargs):
+        super(DuplicateCriteriaInDIDFilter, self).__init__(*args, **kwargs)
+        self._message = 'Duplicate criteria for key/operator in filter expression'
```
Comment (mlassnig) [inline lib/rucio/common/exception.py]: Maybe add the actual duplicate that was found in the expression?

File: lib/rucio/common/utils.py [hunk 3]
```diff
@@ -1081,6 +1083,114 @@ def parse_did_filter_from_string(input_string):
...
+
+    # lookup table mapping operator opposites, used to reverse compound inequalities.
+    operator_opposites_LUT = {
+        'lt': 'gt',
+        'lte': 'gte'
+    }
+    operator_opposites_LUT.update({op2: op1 for op1, op2 in operator_opposites_LUT.items()})
+
+    filters = []
+    if input_string:
```
Comment (mlassnig) [inline lib/rucio/common/utils.py:1148]: I just have to accept that this does what it should :-D

File: lib/rucio/core/did.py [hunk 4]
```diff
@@ -1418,7 +1422,7 @@ def set_status(scope, name, session=None, **kwargs):
 
 @stream_session
 def list_dids(scope, filters, did_type='collection', ignore_case=False, limit=None,
-              offset=None, long=False, recursive=False, session=None):
+              offset=None, long=False, recursive=False, session=None, ignore_dids=None):
```
Comment (mlassnig) [inline lib/rucio/core/did.py]: we try to keep session as the last parameter for consistency

File: lib/rucio/core/did_meta_plugins/filter_engine.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Gabriele Gaetano Fronzé <gfronze@cern.ch>, 2020
+# - Rob Barnsley <rob.barnsley@skao.int>, 2021
+#
+# PY3K COMPATIBLE
+
+from datetime import datetime
```
Comment (mlassnig) [inline lib/rucio/core/did_meta_plugins/filter_engine.py]: correctly sort the imports please

**Review by mlassnig — COMMENTED** [review]
this is amazing.

just two tiny comments.

File: lib/rucio/core/did_meta_plugins/__init__.py [hunk 6]
```diff
@@ -181,11 +181,8 @@ def list_dids(scope=None, filters=None, did_type='collection', ignore_case=False
     :returns: List of dids.
     """
     meta_handler_to_use = None
-
     # Ensure that a single handler manages all the keys of the query
     for key in filters:
-        if key == 'name':
```
Comment (bari12) [inline lib/rucio/core/did_meta_plugins/__init__.py:187]: Why is this removed?

**Review by bari12 — COMMENTED** [review]
Hi @robbarnsley 
This is very nice work. Just one small comment, otherwise this can be added.

We have to think about the next steps though, to me this only gets it "half-way" there. To me the missing steps are:
- Filter engine should not be specific to the column way. Ideally it should prepare a filter which the individual plugin then interpret (or not, if they can't). - So the filter_engine itself should perhaps be sqlalchemy independent.
- Implement it for the column (done) and json plugin
- list_dids should actually be merged (and changed to) list_dids_extended, so that there is only one way of taking care of this. Maybe this can also be part of your next PR.

File: bin/rucio (hunk 1 above)
Comment (robbarnsley) [reply bin/rucio]: done.

File: lib/rucio/common/exception.py (hunk 2 above)
Comment (robbarnsley) [reply lib/rucio/common/exception.py]: done.

File: lib/rucio/core/did.py (hunk 4 above)
Comment (robbarnsley) [reply lib/rucio/core/did.py]: changed.

File: lib/rucio/core/did_meta_plugins/filter_engine.py (hunk 5 above)
Comment (robbarnsley) [reply lib/rucio/core/did_meta_plugins/filter_engine.py]: done.

File: lib/rucio/common/utils.py (hunk 3 above)
Comment (robbarnsley) [reply lib/rucio/common/utils.py:1148]: I appreciate your trust ...

**Comment by bari12** [comment]
Still some conflicts and failing tests, please rebase. But in general I think this is GTG!

**Comment by robbarnsley** [comment]
great, thanks Martin - it's on my list for this week.

**Comment by robbarnsley** [comment]
Sorry took a bit of a while to get around to this, should be OK now.

**Review by bari12 — APPROVED** [review]
